src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -486,44 +487,84 @@ private void ExecuteCompilationActionsCore(ImmutableArray<CompilationAnalyzerAct
             }
         }
 
-        private void ExecuteCodeBlockActionsCore<TLanguageKindEnum>(
-            IEnumerable<CodeBlockStartAnalyzerAction<TLanguageKindEnum>> codeBlockStartActions,
-            IEnumerable<CodeBlockAnalyzerAction> codeBlockActions,
-            IEnumerable<CodeBlockAnalyzerAction> codeBlockEndActions,
-            DiagnosticAnalyzer analyzer,
+        public void ExecuteOperationBlockActions(
+            IEnumerable<OperationBlockStartAnalyzerAction> operationBlockStartActions,
+            IEnumerable<OperationBlockAnalyzerAction> operationBlockActions,
+            IEnumerable<OperationBlockAnalyzerAction> operationBlockEndActions,
+            DiagnosticAnalyzer analyzer, 
             SyntaxNode declaredNode,
             ISymbol declaredSymbol,
-            ImmutableArray<SyntaxNode> executableCodeBlocks,
+            ImmutableArray<IOperation> operationBlocks,
+            ImmutableArray<IOperation> operations,
             SemanticModel semanticModel,
-            Func<SyntaxNode, TLanguageKindEnum> getKind,
-            CodeBlockAnalyzerStateData analyzerStateOpt)
-            where TLanguageKindEnum : struct
+            SyntaxReference declaration,
+            AnalysisScope analysisScope,
+            AnalysisState analysisStateOpt)
         {
+            DeclarationAnalyzerStateData analyzerStateOpt = null;
+
+            try
+            {
+                if (TryStartAnalyzingSyntaxRefence(declaration, analyzer, analysisScope, analysisStateOpt, out analyzerStateOpt))
+                {
+                    ExecuteBlockActionsCore<OperationBlockStartAnalyzerAction, OperationBlockAnalyzerAction, OperationAnalyzerAction, OperationAnalyzerStateData, IOperation, int>(
+                        operationBlockStartActions, operationBlockActions, operationBlockEndActions, analyzer,
+                        declaredNode, declaredSymbol, operationBlocks, (blocks) => operations, semanticModel, null, analyzerStateOpt?.OperationBlockAnalysisState);
+                }
+            }
+            finally
+            {
+                analyzerStateOpt?.ResetToReadyState();
+            }
+        }
+
+        private void ExecuteBlockActionsCore<TBlockStartAction, TBlockAction, TNodeAction, TNodeStateData, TNode, TLanguageKindEnum>(
```

Comment:
At some point in future, we should probably abstract out the common parts of code between:
1. syntax node execution and operation execution
2. code block execution and operation block execution.

I feel a bit worried that a bug fix in one execution path will miss its corresponding counterpart.